### ISSUE TITLE
Missing the login/logout uri in the actuator metrics endpoint

### ISSUE THREAD

**Issue by mohammedalics** [spring-projects/spring-boot#24213]
The filter related endpoints (e.g. /login, /logout) are mapped to uri ROOT, however, it makes sense to be mapped to their URIs and have visibility over their metrics. 

Due to the fact that the spring security handles the logic of the login/logout into a Filter (for instance, login in [UsernamePasswordAuthenticationFilter](https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/authentication/UsernamePasswordAuthenticationFilter.java#L54)), the login request doesn't reach the request handler mapper to have the required attribute (e.g. DATA_REST_PATH_PATTERN_ATTRIBUTE or BEST_MATCHING_PATTERN_ATTRIBUTE) for the actuator to fetch the Uri in [WebMvcTags](https://github.com/spring-projects/spring-boot/blob/e342d7d2626b3c5237d56ac221b0aace8a6a021c/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcTags.java#L152-L156)

To validate the issue, I created this demo repository => https://github.com/mohammedalics/actuator-login-filter-metrics-issue 

** Call Login** 
```
$ curl -kv 'http://localhost:8080/login' -H 'Content-Type: application/x-www-form-urlencoded' --data-raw 'username=user&password=password'`

HTTP/1.1 302
Set-Cookie: JSESSIONID=E0F21CAD780C1C25AAA22DE6F3069764; Path=/; HttpOnly
```

** Call Dummy endpoint**
```
curl -kv 'http://localhost:8080/hello' -H 'Cookie: JSESSIONID=E0F21CAD780C1C25AAA22DE6F3069764'                                                                               

HTTP/1.1 200
hello
```

Then by calling the "http.server.requests" metrics => http://localhost:8080/actuator/metrics/http.server.requests
```

{
  "name": "http.server.requests",
  "description": null,
  "baseUnit": "seconds",
  "measurements": [
    {
      "statistic": "COUNT",
      "value": 4
    },
    {
      "statistic": "TOTAL_TIME",
      "value": 0.22907417300000002
    },
    {
      "statistic": "MAX",
      "value": 0.012099089
    }
  ],
  "availableTags": [
    {
      "tag": "exception",
      "values": [
        "None"
      ]
    },
    {
      "tag": "method",
      "values": [
        "POST",
        "GET"
      ]
    },
    {
      "tag": "uri",
      "values": [
        "REDIRECTION",
        "/hello",
        "root"
      ]
    },
    {
      "tag": "outcome",
      "values": [
        "REDIRECTION",
        "SUCCESS"
      ]
    },
    {
      "tag": "status",
      "values": [
        "302",
        "200"
      ]
    }
  ]
}
```

-- 

If you agree on the issue, I can create a pull request to fix it.

**Comment by wilkinsona** [comment]
> The spring security filter related endpoints (e.g. /login, /logout) are mapped to ROOT uri

I don't think this is the case. The request to `/login` results in the `uri` tag being `REDIRECTION` due to the 302 response that redirects to `/`.

> If you agree on the issue, I can create a pull request to fix it

What sort of fix did you have in mind? Note that when the pattern attributes aren't available the request URL cannot be used as-is as it may result in the `uri` tag's values being unbounded.

**Comment by mohammedalics** [comment]
I understand that in the case of `/login`, the `pattern` [here](https://github.com/spring-projects/spring-boot/blob/e342d7d2626b3c5237d56ac221b0aace8a6a021c/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcTags.java#L113) is `null` due to missing below attributes:
- `DATA_REST_PATH_PATTERN_ATTRIBUTE` which is supposed to be added by RepositoryRestHandlerMapping [here](https://github.com/spring-projects/spring-data-rest/blob/master/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/RepositoryRestHandlerMapping.java#L267)   
- `BEST_MATCHING_PATTERN_ATTRIBUTE` which is supposed to be added by [AbstractUrlHandlerMapping](https://github.com/spring-projects/spring-framework/blob/f1af8a6ae956f2523cb7f2f63710164babac6ddf/spring-webmvc/src/main/java/org/springframework/web/servlet/handler/AbstractUrlHandlerMapping.java#L349) or by RequestMappingInfoHandlerMapping [here](https://github.com/spring-projects/spring-framework/blob/f1af8a6ae956f2523cb7f2f63710164babac6ddf/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/RequestMappingInfoHandlerMapping.java#L172) and [here](https://github.com/spring-projects/spring-framework/blob/f1af8a6ae956f2523cb7f2f63710164babac6ddf/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/RequestMappingInfoHandlerMapping.java#L193)

In this case, the [uri method](https://github.com/spring-projects/spring-boot/blob/e342d7d2626b3c5237d56ac221b0aace8a6a021c/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcTags.java#L111) tries to fulfil the `uri` by using the response and this describes why the request to `/login` results in the `uri` tag being `REDIRECTION`

Those attributes are `RequestMapper` related attributes. If the request doesn't have a controller class and designed to be handled and served in a filter (like the case of `/login`), there will be no chance for the request mapper to add those attributes and then the actuator will not be able to identify the uri (because actuator depends on the request mapper added attributes as mentioned above). 

The fix in my mind is update the [uri method](https://github.com/spring-projects/spring-boot/blob/e342d7d2626b3c5237d56ac221b0aace8a6a021c/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/servlet/WebMvcTags.java#L111) to use the `servletPath` as a kind of fallback if the pattern is null before using the response but use the response if the `servletPath` is just `/` or empty.

**Comment by wilkinsona** [comment]
As I said above, we can't use the servlet path as it would mean that the values of the `uri` tag are potentially unbounded. For example, imagine a request URL where a portion of it is a UUID. If, for whatever reason, there's no pattern attribute available when the `uri` tag is being determined, every request with a different UUID in the URL would result in a different `uri` tag value.

**Comment by mohammedalics** [comment]
Yes, totally make sense. 

Do you think it could be achieved the other way around? I mean spring security adds the `FilterProcessesUrl` to a new attribute that can be caught by the actuator? Maybe we have something already?

**Comment by wilkinsona** [comment]
I don't believe Security configures such an attribute at the moment. @rwinch @jzheaux what are your thoughts on this please?

**Comment by rwinch** [comment]
I'm curious how we would generalize this to all of Spring Security's `Filter`s or even more generally, any `Filter`. The challenge is that many `Filter`s are involved in processing a request while only one MVC endpoint is involved. For example, during a typical form based log in at minimum the following `Filter`s are going to be used: `HeaderWriterFilter`, `CsrfFilter`, and `UsernamePasswordAuthenticationFilter`. It doesn't seem that we can set a single attribute here. W It seems that MVC was already setting the attribute and Metrics leveraged it. It seems strange that Security would set an attribute specifically for metrics. Perhaps we can provide hooks that allow Metrics to set attributes though?

To be clear I'm not saying that we cannot get something to work, but just that I'm not sure how we would expect it to work. This may be obvious to someone more experienced with metrics, but I'm very new to this space.

**Comment by philwebb** [comment]
See also #2563

**Comment by marwin1991** [comment]
I think it is sill and issue :)
